packages/contracts/src/contracts/current/protocol/AssetTransferProxy/AssetTransferProxy.sol

```diff
@@ -0,0 +1,116 @@
+/*
+
+  Copyright 2018 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.4.21;
+
+import "./IAssetTransferProxy.sol";
+import "./IAssetProxy.sol";
+import "../../utils/Authorizable/Authorizable.sol";
+
+contract AssetTransferProxy is
+    Authorizable,
+    IAssetTransferProxy
+{
+    // Mapping from Asset Proxy Id's to their respective Asset Proxy
+    mapping (uint8 => IAssetProxy) public assetProxies;
+
+    /// @dev Delegates transfer to the corresponding asset proxy.
+    /// @param assetMetadata Byte array encoded for the respective asset proxy.
+    /// @param from Address to transfer token from.
+    /// @param to Address to transfer token to.
+    /// @param amount Amount of token to transfer.
+    function transferFrom(
+        bytes assetMetadata,
+        address from,
+        address to,
+        uint256 amount)
+        public
+        onlyAuthorized
+    {
+        // Lookup asset proxy
+        require(assetMetadata.length >= 1);
+        uint8 assetProxyId = uint8(assetMetadata[0]);
+        IAssetProxy assetProxy = assetProxies[assetProxyId];
+        require(assetProxy != address(0x0));
+
+        // Delegate transfer to asset proxy
+        assetProxy.transferFrom(assetMetadata, from, to, amount);
+    }
+
+    /// @dev Registers a new asset proxy.
+    /// @param assetProxyId Id of the asset proxy.
+    /// @param newAssetProxyAddress Address of the asset proxy contract to register.
+    /// @param currentAssetProxyAddress Address of existing asset proxy to overwrite.
+    function registerAssetProxy(
+        uint8 assetProxyId,
```

Comment:
There's an attack vector here: the owner of this contract can deregister a proxy and then register a new proxy to the same ID, causing existing orders to be valid for a different proxy. To avoid this, we can store a nonce that increments every time a new proxy is stored, and then uses the nonce as the new proxy's ID. This should guarantee that no ID is every reused.

Reply:
We chatted about this offline and decided to keep the current implementation for the time being. Ultimately, this is more of a trust issue than a security vulnerability. Two events would need to take place for such an attack to play out: (i) The contract owner would have to swap proxies, and (ii) the enduser would have to set allowances on both proxies. The assumptions that we are not a bad actor and the enduser knows what they're signing are fundamental to our protocol. If either of these failed, then our entire system would be vulnerable.